include/clap/ext/draft/context-menu.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+
+#include "../../plugin.h"
+
+// This extension lets the host and plugin exchange menu items and let the plugin ask the host to
+// show its context menu.
+
+static CLAP_CONSTEXPR const char CLAP_EXT_CONTEXT_MENU[] = "clap.context-menu.draft/0";
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// There can be different target kind for a context menu
+enum {
+   CLAP_CONTEXT_MENU_TARGET_KIND_GLOBAL = 0,
+   CLAP_CONTEXT_MENU_TARGET_KIND_PARAM = 1,
+};
+
+// Describes the context menu target
+typedef struct clap_context_menu_target
+{
+   uint32_t kind;
+   clap_id id;
+} clap_context_menu_target_t;
+
+// Flags for the context menu accelerator
+enum {
+   CLAP_CONTEXT_MENU_ACCELERATOR_MODIFIER_SHIFT = (1 << 0),
+   CLAP_CONTEXT_MENU_ACCELERATOR_MODIFIER_CONTROL = (1 << 1),
+   CLAP_CONTEXT_MENU_ACCELERATOR_MODIFIER_ALT = (1 << 2),
+
+   // Windows key or Command key
+   CLAP_CONTEXT_MENU_ACCELERATOR_MODIFIER_META1 = (1 << 3),
+};
+
+// Describes a single key strike with its modifiers
+typedef struct clap_context_menu_accelerator_key {
+   uint32_t flags;
+   uint32_t key; // ASCII key
+} clap_context_menu_accelerator_key_t;
+
+// Describes an accelerator which can consist of a key strike combo
+typedef struct clap_context_menu_accelerator {
+   uint32_t combo_size;
+   clap_context_menu_accelerator_key_t keys[4];
+} clap_context_menu_accelerator_t;
+
+// Context menu builder
+typedef struct clap_context_menu_builder {
+   void *ctx;
+
+   // accelerator may be null
+   bool(CLAP_ABI *add_entry)(const struct clap_context_menu_builder *builder,
+                             const char                             *label,
+                             const clap_context_menu_accelerator_t  *accelerator,
+                             bool                                    is_enabled,
+                             clap_id                                 action_id);
+
+   // accelerator may be null
+   bool(CLAP_ABI *add_checkmark)(const struct clap_context_menu_builder *builder,
+                                 const char                             *label,
+                                 const clap_context_menu_accelerator_t  *accelerator,
+                                 bool                                    is_enabled,
+                                 bool                                    is_checked,
+                                 clap_id                                 action_id);
+
+   bool(CLAP_ABI *add_separator)(const struct clap_context_menu_builder *builder);
+
```

Comment:
Can this API be made more future proof (and simpler) by using flags? I.e.:

```c
typedef enum clap_context_menu_flags {
  CLAP_CONTEXT_MENU_ENTRY_DISABLED = 1,
  CLAP_CONTEXT_MENU_ENTRY_CHECKBOX = 2,
  CLAP_CONTEXT_MENU_ENTRY_CHECKED = 4,
  CLAP_CONTEXT_MENU_ENTRY_SEPARATOR = 8,
} clap_context_menu_flags_t;
bool (CLAP_ABI *add_entry) (const struct clap_context_menu_builder *builder,
                            const char                             *label,
                            clap_context_menu_flags_t               flags,
                            clap_id                                 action_id);
```

That allows adding _TITLE, _RADIO, and other menu item types in the future and uses 1 add-entry function instead of three.

Reply:
My gut reaction is that this would impact performance, and that it gives extra runtime flexibility that most people won't need, but maybe that's OK? If I understand correctly, it only gets run a few times, and only ever in the main thread.

Reply:
Your proposal as it is, doesn't work because it doesn't provide the checked state.
Different menu entries requires different data.

I think this menu builder is quite simple, but can probably do the job for the next 10 years.

Reply:
> Your proposal as it is, doesn't work because it doesn't provide the checked state. Different menu entries requires different data.
> 
> I think this menu builder is quite simple, but can probably do the job for the next 10 years.

The checked state is flag 0x04.

Reply:
So what are the other entries that you'd like to add?

Reply:
The menu builder with methods looks a lot cleaner to me than flags. I think if one has to implement the methods, it doesn't leave so much  room for interpretation/bad combinations of flags. Flags would provide some data compression but only for a short moment during runtime. Most of the time one would probably directly unpack them again as bools.

But I also wonder a bit how a "radio" element would be added in the future. Maybe as a special kind of submenu where each entry must be a check entry which is then checked exclusively? Otherwise for a special radio element with an index value, perform would have to be called with an int value.

Reply:
> So what are the other entries that you'd like to add?

Personally, I don't need any atm. ;-)

I've just designed and seen enough menu APIs in the past to have an idea what feature requests might come up over time. I've outlined some types above already, example:

![image](https://user-images.githubusercontent.com/281887/207918080-d3e2ac76-debc-46d5-86c9-28b81bff244f.png)

This could be described by an array of:
```c
struct Entry {
  uint_t item_type; // TITLE, SEPARATOR, ITEM, RADIO, CHECKBOX
  clap_id_t id;           // identifier for activation
  uint64_t flags;       // DISABLED=0x1, CHECKED=0x2
  const char *const label; // unused for SEPARATOR
  const char *const icon; // unicode char, or "mi-<materialicon>", "fa-<fontawesomeicon>", etc
  const char *const hotkey; // just a hint, ignored by most hosts to avoid conflicts
  const char *const category; // either empty or the label of a menu item that is a submenu parent
  const char *const tooltip; // brief description (might need more than 256 chars in procedural API)
};
// Consecutive radio items form a group that has a single unique selection 
```

In a declarative API, the returned array and `char*` strings are assumed to be static const, i.e. persist while the plugin is loaded. The `category` label is a bit awkward for long lists of menu items, but it allows to keep all items in a flat declarative array. If a procedural API is chosen like yours, `begin_submenu() + end_submenu()` works fine, and of course the strings would be `char label[256];` instead of static const.
 
Don't expect all hosts to make use of the submenu facility though, e.g. Anklang will present items in a long vertically wrapped list that is horizontally scrollable and not support submenus any time soon. Handling large submenus  can easily get out of hand and be confusing for users, especially if created via generic APIs, rather than designed by a UI expert.
Note that a declarative API that returns a single array can make things much simpler (and faster) if long lists of menu items are used (e.g. for bank or preset selections). Especially if checked/disabled states are encoded with the rest of the menu entry information, because that essentially requires re-querying of the whole menu hierarchy for every popup, just to update the DISABLED and CHECKED booleans.

Reply:
One of the original proposals had the plugin provide an array to the host. The downsides of doing it that way is that the plugin also needs to allocate storage for the array. With the builder style an intermediate data structure isn't needed. If the host represents the context menu as some kind of tree, the builder can immediately construct that tree.

Reply:
> Don't expect all hosts to make use of the submenu facility though, e.g. Anklang will present items in a long vertically wrapped list that is horizontally scrollable and not support submenus any time soon. Handling large submenus can easily get out of hand and be confusing for users, especially if created via generic APIs, rather than designed by a UI expert.

I think the current extension proposal (with submenus) should work fine for everybody. I mean, everyone can interpret the submenus as desired. E.g. Anklang is totally free to put the submenus on the top level as well - maybe surrounded with separators and using the submenu labels as section titles.
Otherwise there could be a flag or method that tells if a menu builder supports submenus. If such a concept is established, maybe there also could be other flags in the future like "supports radio entries".

Reply:
I've reworked the context menu to have a design that is more "future proof".

Regarding the icon, I like the option to use unicode chars, but otherwise I don't want to push requirements on the host to provide thousand of standard icons. I don't have enough knowledge with that to take a good decision.

hotkey, we decided to get rid of it to avoid confusion about if is was a way to bind a shortcut to the host.

Reply:
If we are happy with the design, I'd like then to move forward.

This extension will be published as draft, so we can still improve and add more data to the menu entries.

If someone has enough knowledge and wants to make a proposition, then please make a PR and consider what the host or plugin would have to do to ensure the availability of icons/unicode characters.